[GStreamer] Video sink should discard late frames to avoid audio and video desynchronization

calvaris opened the pull request:
#### 1302f18a970637284e926d350a0ab853d0cf9a15
<pre>
[GStreamer] Video sink should discard late frames to avoid audio and video desynchronization
<a href="https://bugs.webkit.org/show_bug.cgi?id=247740">https://bugs.webkit.org/show_bug.cgi?id=247740</a>

Reviewed by Philippe Normand.

In some cases and/or platforms video frames can arrive late and if are not discarded, they end up being painted and
there can be a/v desynchronization. Setting the max-lateness property forces the sink to discard late frames and solve
the problem.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::createVideoSink):

Canonical link: <a href="https://commits.webkit.org/256650@main">https://commits.webkit.org/256650@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1958c20af4f056b3d6d5aa123cfc7e57ad73ef7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96418 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5672 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29486 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105961 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166310 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100401 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5847 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34419 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88797 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102687 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102089 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83020 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31336 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74229 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40148 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37822 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20970 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4612 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/3844 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/876 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40240 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3928,6 +3928,10 @@ GstElement* MediaPlayerPrivateGStreamer::createVideoSink()
 
+    // In some context, we could end up with a/v desync so we need to ensure late frames are discarded.
+    uint64_t maxLateness = 100 * GST_MSECOND;
+    g_object_set(m_videoSink.get(), "max-lateness", maxLateness, nullptr);
```
is `m_videoSink` a basesink sub-class in all cases? At least our custom GL sink isn't, it's a bin, so setting max-lateness in that situation will cause havoc :)

calvaris replied:
If the property does not exist, it should be harmless other than maybe getting a warning. I see two options here:
1. Adding the property to all individual sinks.
2. Adding the property here but checking for special sinks like the GL one.

Fun fact: I had the code just for the GL sink and decided to move it there to make it more general :sweat_smile:

philn replied:
Well yes, since no check is done in this current patch, i don't see how any issue would be solved unless you're lucky at runtime and you get a basesink sub-class.

So can you check all the cases we support (GL, non-GL, punch-hole) and also don't do this if the source is a MediaStream please, because sync is disabled for that case anyway.

calvaris replied:
I think I solved this.

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3841,6 +3841,26 @@ GstElement* MediaPlayerPrivateGStreamer::createVideoSink()
     acceleratedRenderingStateChanged();
 
+    // Ensure the sink has the max-lateness property set.
+    auto exit = makeScopeExit([this]() {
```
I think you can omit parenthesis and use &, `[&] {` maybe?

calvaris replied:
I removed the () but I prefer to keep the this because I prefer to control the variables I capture when I write lambdas.

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3841,6 +3841,26 @@ GstElement* MediaPlayerPrivateGStreamer::createVideoSink()
+        GstElement* sink = m_videoSink.get();
+        while (GST_IS_BIN(sink)) {
+            GUniquePtr<GstIterator> iter(gst_bin_iterate_sinks(GST_BIN_CAST(sink)));
+            GValue returnValue = G_VALUE_INIT;
```
maybe rename to `value`?

philn reviewed (approved):
LGTM, only a couple nits for your consideration :)

calvaris committed:
[GStreamer] Video sink should discard late frames to avoid audio and video desynchronization
https://bugs.webkit.org/show_bug.cgi?id=247740

Reviewed by Philippe Normand.

In some cases and/or platforms video frames can arrive late and if are not discarded, they end up being painted and
there can be a/v desynchronization. Setting the max-lateness property forces the sink to discard late frames and solve
the problem.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::createVideoSink):

Canonical link: https://commits.webkit.org/256650@main
